### ISSUE TITLE
Recommended tutorial to base an Re-Identification project on?

### ISSUE THREAD

**Issue by joaqo** [KevinMusgrave/pytorch-metric-learning#284]
Hi, I've been trying out the project for some days now, really like it a lot. The code is really well structured, and I also found record-keeper to be brilliant. Congrats on the great project!

I've been trying the samples and I'm having a hard time deciding on a good baseline to start working on a simple Vehicle Re-Identification project. Any recommendations? Simplicity over final precision would be preferred. The dataset would probably have thousands of different vehicle ids (which would translate to classes if I ended up using a cross-entropy loss).

Also, in the search for this baseline I've been training the samples provided in the project. I expanded the training for more epochs in the [ClassDisjointCIFAR100 metric loss sample](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/examples/notebooks/MetricLossOnly.ipynb) and added some TensorBoard metrics and got the following results:

<img width="1386" alt="image" src="https://user-images.githubusercontent.com/14966348/109519565-1d43f300-7a8a-11eb-84f9-b9bbaaa91b4b.png">

They seemed a bit low to me, are these results to be expected or is it possible I am doing something wrong?

Thanks!

**Comment by KevinMusgrave** [comment]
> Hi, I've been trying out the project for some days now, really like it a lot. The code is really well structured, and I also found record-keeper to be brilliant. Congrats on the great project!

Thanks 😄 

> I've been trying the samples and I'm having a hard time deciding on a good baseline to start working on a simple Vehicle Re-Identification project. Any recommendations? Simplicity over final precision would be preferred. The dataset would probably have thousands of different vehicle ids (which would translate to classes if I ended up using a cross-entropy loss).

Though this library has some losses from re-id papers (ArcFace, CosFace etc.), I'm not really familiar with the field to know what the best settings are. If you're dealing with thousands of classes, then memory might become an issue with classification based losses like cross-entropy, ArcFace, and CosFace.

> Also, in the search for this baseline I've been training the samples provided in the project. I expanded the training for more epochs in the [ClassDisjointCIFAR100 metric loss sample](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/examples/notebooks/MetricLossOnly.ipynb) and added some TensorBoard metrics and got the following results:
> 
> <img alt="image" width="1386" src="https://user-images.githubusercontent.com/14966348/109519565-1d43f300-7a8a-11eb-84f9-b9bbaaa91b4b.png">
> 
> They seemed a bit low to me, are these results to be expected or is it possible I am doing something wrong?

It would probably be better to have examples that showed great results, but I didn't get around to tuning the parameters, and I actually never ran the examples for as long as you did. 

Regarding baselines/frameworks, this repo might be helpful for you: https://github.com/KaiyangZhou/deep-person-reid

It's for person re-id, but I assume that it can be applied to vehicle re-id. If you want to use pytorch-metric-learning losses/miners, they can probably be imported into that framework.

**Comment by KevinMusgrave** [comment]
Actually the numbers you got with the example notebook do seem a bit low. I'm training for 20 epochs and I'll post the results here when it's done.

**Comment by KevinMusgrave** [comment]
Here's what I got after 20 epochs. It's not great, not terrible.
<img width="1292" alt="metricLossOnly20Epochs" src="https://user-images.githubusercontent.com/2314891/109676054-56419d80-7b46-11eb-95b5-f9fb60cc2571.png">
<img width="874" alt="metricLossOnly20EpochsUMAP" src="https://user-images.githubusercontent.com/2314891/109676066-58a3f780-7b46-11eb-9a15-ea82eb4b59bc.png">

**Comment by joaqo** [comment]
I think the issue was just specific to the `ClassDisjointCIFAR100 metric loss` sample. I am now working on a custom training script based on `MNIST using TripletMarginLoss` modified to use CIFAR100 (though not disjointed) and a basic resnet-50 network and I'm getting a pretty good 0.7 precision_at_1 with a much less noisy graph.

<img width="1365" alt="image" src="https://user-images.githubusercontent.com/14966348/109670854-4b8d0580-7b52-11eb-9d6e-73b6796902e0.png">

Is this on the ballpark of what you would expect or do you think it could be improved?

My objective at the moment is to create the simplest re-id script I can, that has decent accuracy, and works well on several re-id datasets. The distinctive feature of these types of datasets is that they have an extremely large amount of classes, with each class having a small amount of images. If I get good results it could end up as a PR to this repo if you are interested.

Regarding deep-person-reid, I worked with it for some time but found it to be overtly complex. Training on the supported datasets works well, but when I tried training on custom data I had issues. A combination of failing to overfit a small custom 300 image dataset, and reading your awesome reality check paper convinced me to go back to basics and try to create the simplest re-id training script I could. Using this approach I have already overfitted the dataset I had previously failed to, so I'm just going to continue working in this simple way.

At the moment I am trying to overfit a 3000 class / 38k image dataset using TripletMarginLoss + CosineSimilarity, will post results. If you have any suggestions on things I could try or any ideas they are more than welcome!

**Comment by joaqo** [comment]
Oh just saw your last post, you're getting much better results than I am, interesting. Is this just the plain 'ClassDisjointCIFAR100 metric loss' uploaded to jupyter set to run for more epochs?

**Comment by KevinMusgrave** [comment]
> Oh just saw your last post, you're getting much better results than I am, interesting. Is this just the plain 'ClassDisjointCIFAR100 metric loss' uploaded to jupyter set to run for more epochs?

Yes. Did you run the notebook in google colab? If not, what version of pytorch and pytorch-metric-learning do you have installed?

> I think the issue was just specific to the `ClassDisjointCIFAR100 metric loss` sample. I am now working on a custom training script based on `MNIST using TripletMarginLoss` modified to use CIFAR100 (though not disjointed) and a basic resnet-50 network and I'm getting a pretty good 0.7 precision_at_1 with a much less noisy graph.
> 
> <img alt="image" width="1365" src="https://user-images.githubusercontent.com/14966348/109670854-4b8d0580-7b52-11eb-9d6e-73b6796902e0.png">
> 
> Is this on the ballpark of what you would expect or do you think it could be improved?

Looks reasonable to me. Be careful with the ```k``` value in AccuracyCalculator. I'm assuming you left it as ```k=1``` which is why all the metrics look identical to precision at 1. If you're only interested in precision at 1, you can do: 
```AccuracyCalculator(include=("precision_at_1",), k=1)```

> My objective at the moment is to create the simplest re-id script I can, that has decent accuracy, and works well on several re-id datasets. The distinctive feature of these types of datasets is that they have an extremely large amount of classes, with each class having a small amount of images. If I get good results it could end up as a PR to this repo if you are interested.

That would be great! I think it would be very helpful to have more examples in this repo that use something other than CIFAR and MNIST. Currently there's [one notebook by wconnell](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/examples/notebooks/scRNAseq_MetricEmbedding.ipynb).

> Regarding deep-person-reid, I worked with it for some time but found it to be overtly complex. Training on the supported datasets works well, but when I tried training on custom data I had issues. A combination of failing to overfit a small custom 300 image dataset, and reading your awesome reality check paper convinced me to go back to basics and try to create the simplest re-id training script I could. Using this approach I have already overfitted the dataset I had previously failed to, so I'm just going to continue working in this simple way.

> At the moment I am trying to overfit a 3000 class / 38k image dataset using TripletMarginLoss + CosineSimilarity, will post results. If you have any suggestions on things I could try or any ideas they are more than welcome!

Sounds good. My only suggestion is that I think ContrastiveLoss usually works better than TripletMarginLoss.

**Comment by joaqo** [comment]
Tried it on a clean colab and I'm now getting the same results you are, I think originally I was trying too many changes and I must have broken something in the process.

Those are some great tips! I'm gonna keep trying stuff out and report back if I get anything interesting, thanks!

**Comment by joaqo** [comment]
Hey @KevinMusgrave I'm about to run some training jobs on some large vehicle-reid datasets and was wondering if you had any specific recommendations on how to use ContrastiveLoss. I noticed you recommended me using it but I couldn't find any samples showing it in use.

These are the relevant lines I am thinking of changing:

```python
distance = distances.CosineSimilarity()
reducer = reducers.ThresholdReducer(low=0)
loss_func = losses.TripletMarginLoss(margin=0.2, distance=distance, reducer=reducer)
mining_func = miners.TripletMarginMiner(
    margin=0.2, distance=distance, type_of_triplets="semihard"
)
accuracy_calculator = AccuracyCalculator()
```

I was going to just set `loss_func = ContrastiveLoss()`, but I figured you might have some recommendations on how to set the other operations to match it. Thanks!

**Comment by KevinMusgrave** [comment]
If you're aiming for simplicity, I'd get rid of the other stuff and use just ContrastiveLoss(). For the best results, you'll need to try out different values for its hyperparameters, [```pos_margin``` and ```neg_margin```](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#contrastiveloss)

**Comment by joaqo** [comment]
Gotcha, running some jobs to compare my previous setup vs ContrastiveLoss. Will tune the margins as I get results. Thanks!

**Comment by joaqo** [comment]
Hi @KevinMusgrave I wanted to ask you a pretty off-topic question about your paper, and didn't want to pollute the repo with a new issue for it, so I thought I'd just ask here. Please feel free to completely ignore this comment.

In your paper, when comparing metrics you present these images:
<img width="1202" alt="image" src="https://user-images.githubusercontent.com/14966348/114591021-d1c35e00-9c5f-11eb-9dfa-1ffb1b3bb9d9.png">

And mention that they all have Recall@1 of around 100%. If I understand Recall@k correctly this would mean that there is only one positive example in the whole dataset right? Perfect recall means that you have captured all possible matches in the dataset, and if you achieve perfect recall on your first and only prediction (k=1), that would mean there is only one match to make in the whole dataset? Nevertheless, in the images there are several positive/negative samples, so I am clearly misunderstanding something.

I'd be extremely grateful if you could expand a bit on what I am missing here.

Thanks again!

**Comment by KevinMusgrave** [comment]
I think this issue should help: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/283

**Comment by joaqo** [comment]
Thats perfect, thanks!
